### DISCUSSION TITLE
Avoid auto-importing 'greenlet'

### DISCUSSION THREAD

**Discussion by stianjensen** [celery/celery#8169] [General]
We're running celery in production as part of a django-application. Celery is therefore imported as part of our django-workers when running under wsgi, to be able to queue celery tasks.

Unrelated to this setup, we've installed bpython in the same environment, as a convenience to have a shell readily available for debugging. Installing bpython also installs the package [greenlet](https://pypi.org/project/greenlet/).

Without us explicitly importing either bpython or greenlet in our code, it turns out that celery has suddenly auto-imported greenlet for us:
https://github.com/celery/celery/blob/ed71ebb2addd0579e2f64e15e9d44ec7a1e31434/celery/utils/threads.py#L13

We recently discovered this, when investigating a crash in production after greenlet was updated to its new version 2.x

Is it correct that celery should just implicitly auto-import this package when it is present (for some unrelated reason) in the enviroment?

**Comment by auvipy** [comment]
AFAIK https://docs.celeryq.dev/en/stable/userguide/workers.html#concurrency greenlet should not be auto imported. also the version of greenlet should be version fixed with specific release

**Reply by stianjensen** [reply]
I'm currently running celery 5.2.7 with Redis. For debugging I've added an exception inside greenlet's bootstrapper:
Then, when I but my app's webserver, I get the following stacktrace (simplified for readability):
```
File "/app/hyre/__init__.py", line 1, in <module>
    from hyre.celery import app as celery_app
  File "/app/hyre/celery.py", line 7, in <module>
    from celery import Celery
  File "<frozen importlib._bootstrap>", line 1075, in _handle_fromlist
  File "/app/.venv/lib/python3.10/site-packages/celery/local.py", line 469, in __getattr__
    module = __import__(self._object_origins[name], None, None,
  File "/app/.venv/lib/python3.10/site-packages/celery/app/__init__.py", line 2, in <module>
    from celery import _state
  File "/app/.venv/lib/python3.10/site-packages/celery/_state.py", line 15, in <module>
    from celery.utils.threads import LocalStack
  File "/app/.venv/lib/python3.10/site-packages/celery/utils/threads.py", line 13, in <module>
    from greenlet import getcurrent as get_ident
  File "/app/.venv/lib/python3.10/site-packages/greenlet/__init__.py", line 9, in <module>
```

I would expect greenlet to not be imported at all since I have not configured it.
The reason greenlet is not pinned is that I haven't asked celery to install greenlet at all (it came from another unrelated dependency).

Following the code through the stacktrace now, I find no attempt at reading configuration before attempting to import the module 🤔

**Reply by stianjensen** [reply]
It looks like it's been happening there for 10+ years:
https://github.com/celery/celery/commit/41d90c9f8af15469ed1db8531f8902baf187a70a

**Reply by auvipy** [reply]
since it is a very old issue, we have to change it very carefully. what do you think could be a right approach to get rid of this?

**Reply by stianjensen** [reply]
Yeah, I guess it may not be easy. Maybe it will have to wait for a potential 6.x to be conservative.

A new opt-in flag to *disable* this behavior could help, but is a kind of strange configuration variable to introduce.

I'm not familiar enough with the concurrency modes of celery to understand in what way this import has affected _celery_ behavior, and therefore to what extent a fix that stops auto-importing it will be breaking, but in the same way that this bug _did_ affect our code-base even though we weren't really depending on greenlet at all, I guess removing it _could_ also be breaking for others that aren't even aware of it.

**Reply by auvipy** [reply]
true indeed
